unstable/propertyvalue/propertyvalue.go

```diff
@@ -108,6 +110,81 @@ func TransformPropertyValue(
 	return transformer(path, value)
 }
 
+type LimitDescentError struct{}
+
+func (LimitDescentError) Error() string {
+	return "limit descent"
+}
+
+// TransformPropertyValueLimitDescent is a variant of TransformPropertyValue that allows the transformer
+// to return a LimitDescentError to indicate that the recursion should not descend into the value without
+// aborting the whole transformation.
+func TransformPropertyValueLimitDescent(
+	path resource.PropertyPath,
+	transformer func(resource.PropertyPath, resource.PropertyValue) (resource.PropertyValue, error),
+	value resource.PropertyValue,
+) (resource.PropertyValue, error) {
```

Comment:
This function conflates a bunch of stuff. Traversal order is not related to supporting a `LimitDescentError`. The idiomatic Go API for tree walking looks like this:

```go
var SkipChildren = skipChildren{}

type Transformer = func(
	resource.PropertyPath, resource.PropertyValue, entering bool,
) (resource.PropertyValue, error)

func TransformPropertyValue(
	path resource.PropertyPath,
	transformer Transformer,
	value resource.PropertyValue,
) (resource.PropertyValue, error)
```

We can then use `TransformPropertyValue` to construct all the functions we need:

```go
func TransformPropertyValuePreorder(
	path resource.PropertyPath,
	transformer func(
		resource.PropertyPath, resource.PropertyValue,
	) (resource.PropertyValue, error),
	value resource.PropertyValue,
) (resource.PropertyValue, error) {
	return TransformPropertyValue(path, func(
		path resource.PropertyPath, value resource.PropertyValue, entering bool,
	) (resource.PropertyValue, error) {
		if entering { return transformer(path, value) } 
		return value, nil 
	}, value)
}
```

Can we keep to a general function to handle the traversal?

Reply:
>Traversal order is not related to supporting a LimitDescentError

Not sure I understand what you mean here - if you visit children before their parents then a visitor doesn't get a chance to return a `LimitDescentError`.

The API you suggested is cleaner and I have indeed seen the pattern with `entering` elsewhere, so happy to refactor it like that.

Reply:
Refactored the module and kept just one traversal function. LMK what you think!

Reply:
Yeah I recall seeing `entering` somewhere but I can't recall where. This latest version is OK-ish to reason about. The boolean needs to be called `entering` otherwise the API causes boolean blindness.

Do you have refs to Go libs where you've seen this?

I'd suspect idiomatic Go would be further reliant on mutation :) But there's e.g. https://github.com/hashicorp/go-cty/blob/v1.4.1-0.20200414143053-d3edf31b6320/cty/walk.go#L87 that does not mutate.

Reply:
I've just found https://github.com/fogfish/golem for light weekend reading, this is going to be fun but we probably don't want to go too far down that route.

Reply:
https://github.com/fogfish/golem this looks pretty fun indeed

Reply:
https://github.com/pulumi/pulumi-terraform-bridge/blob/f02a167db6fded877fb560cf52b8815817e95f2d/pkg/tfgen/parse/extension.go#L118

The Markdown parsing library we use uses the `entering` pattern for its recursion.